protocol/msgpack/aggregated_encoder.go

```diff
@@ -0,0 +1,111 @@
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"github.com/m3db/m3metrics/metric/aggregated"
+	"github.com/m3db/m3metrics/policy"
+)
+
+type encodeRawMetricWithPolicyFn func(data []byte, p policy.Policy)
+type encodeRawMetricFn func(data []byte)
+type encodeMetricAsRawFn func(m aggregated.Metric) []byte
+
+// aggregatedEncoder uses MessagePack for encoding aggregated metrics.
+// It is not thread-safe.
+type aggregatedEncoder struct {
+	encoderBase
+
+	buf                         encoderBase                 // buffer for encoding metrics
+	encodeRootObjectFn          encodeRootObjectFn          // root object encoding function
+	encodeRawMetricWithPolicyFn encodeRawMetricWithPolicyFn // raw metric with policy encoding function
+	encodeRawMetricFn           encodeRawMetricFn           // raw metric encoding function
+	encodeMetricAsRawFn         encodeMetricAsRawFn         // metric to raw metric conversion function
+}
+
+// NewAggregatedEncoder creates an aggregated encoder
+func NewAggregatedEncoder(encoder BufferedEncoder) AggregatedEncoder {
+	enc := &aggregatedEncoder{
+		encoderBase: newBaseEncoder(encoder),
+		buf:         newBaseEncoder(newBufferedEncoder()),
+	}
+
+	enc.encodeRootObjectFn = enc.encodeRootObject
+	enc.encodeRawMetricWithPolicyFn = enc.encodeRawMetricWithPolicy
+	enc.encodeRawMetricFn = enc.encodeRawMetric
+	enc.encodeMetricAsRawFn = enc.encodeMetricAsRaw
+
+	return enc
+}
+
+func (enc *aggregatedEncoder) Encoder() BufferedEncoder      { return enc.encoder() }
+func (enc *aggregatedEncoder) Reset(encoder BufferedEncoder) { enc.reset(encoder) }
+
+// NB(xichen): we encode metric as a raw metric so the decoder can inspect the encoded byte stream
+// and apply filters to the encode bytes as needed without fully decoding the entire payload
+func (enc *aggregatedEncoder) EncodeMetricWithPolicy(mp aggregated.MetricWithPolicy) error {
+	if err := enc.err(); err != nil {
+		return err
+	}
+	enc.encodeRootObjectFn(rawMetricWithPolicyType)
+	data := enc.encodeMetricAsRawFn(mp.Metric)
+	enc.encodeRawMetricWithPolicyFn(data, mp.Policy)
+	return enc.err()
+}
+
+func (enc *aggregatedEncoder) EncodeRawMetricWithPolicy(rp aggregated.RawMetricWithPolicy) error {
+	if err := enc.err(); err != nil {
+		return err
+	}
+	enc.encodeRootObjectFn(rawMetricWithPolicyType)
+	enc.encodeRawMetricWithPolicyFn(rp.RawMetric.Bytes(), rp.Policy)
+	return enc.err()
+}
+
+func (enc *aggregatedEncoder) encodeRootObject(objType objectType) {
+	enc.encodeVersion(aggregatedVersion)
+	enc.encodeNumObjectFields(numFieldsForType(rootObjectType))
+	enc.encodeObjectType(objType)
+}
+
+// NB(xichen): we do not encode the number of object fields here because the metric
+// is encoded as raw bytes in the payload, which means each metric is decoded individually
+// rather than in a streaming fashion. As a result we can safely make backward-compatible
+// changes (e.g., encode one more field at the end) without needing the number of object
+// fields.
```

Comment:
Might it be worth encoding the number of fields anyway for consistency?  Also if anyone wants to implement a streaming decoder this would be most helpful.  It would only be a single varint.